File: solution/107BinaryTreeLevelOrderTraversalII/107BinaryTreeLevelOrderTraversalII.go
package main

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Node struct {
	t *TreeNode
	l int
}

type Queue []Node

func (q *Queue) Empty() bool { return len(*q) == 0 }
func (q *Queue) Push(n Node) { *q = append(*q, n) }
func (q *Queue) Pop() (n Node) {
	n = (*q)[0]
	*q = (*q)[1:]
	return
}

func reverse(numbers [][]int) [][]int {
	for i := 0; i < len(numbers)/2; i++ {
		j := len(numbers) - i - 1
		numbers[i], numbers[j] = numbers[j], numbers[i]
	}
	return numbers
}

func levelOrderBottom(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}
	var results [][]int
	var result []int
	var q Queue
	q.Push(Node{root, 0})
	last := -1
	for !q.Empty() {
		p := q.Pop()
		if p.l != last {
			if result != nil {
				results = append(results, result)
			}
			last = p.l
			result = nil
		}
		result = append(result, p.t.Val)
		if p.t.Left != nil {
			q.Push(Node{p.t.Left, p.l + 1})
		}
		if p.t.Right != nil {
			q.Push(Node{p.t.Right, p.l + 1})
		}
	}
	if result != nil {
		results = append(results, result)
	}
	return reverse(results)
}

func main() {

}

File: solution/107BinaryTreeLevelOrderTraversalII/107BinaryTreeLevelOrderTraversalII.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn level_order_bottom(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut queue = std::collections::VecDeque::new();
        let mut last_level:i32 = -1;
        if let Some(node) = root {
            queue.push_back((node, 0));
        }
        let mut results = vec![];
        while let Some((node, level)) = queue.pop_front() {
            if last_level != level as i32 {
                results.push(vec![]);
                last_level = level as i32;
            }
            let mut n = node.borrow_mut();
            results[level].push(n.val);
            if n.left.is_some() {
                queue.push_back((Rc::clone(n.left.as_mut().unwrap()), level+1));
            }
            if n.right.is_some() {
                queue.push_back((Rc::clone(n.right.as_mut().unwrap()), level+1));
            }
        }
        results.into_iter().rev().collect()
    }
}
